src/tokens/mod.rs

```diff
@@ -0,0 +1,654 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! Provides a simple stream API bi-directionally over [`IonReader`](crate::IonReader).
+//!
+//! Conceptually [`TokenStream`] can be thought of as a continuation of the computation of
+//! an Ion data stream.  This is useful for composing and transforming over streams of Ion data.
+//!
+//! It pulls in parts of the [element crate](crate::element) API to make it easier to work
+//! with values without pulling in fully materializing the tree.
+
+use crate::element::{Annotations, Bytes, Value};
+use crate::result::illegal_operation;
+use crate::text::text_formatter::IonValueFormatter;
+use crate::thunk::{Thunk, ThunkState};
+use crate::{Decimal, Int, IonError, IonResult, IonType, Str, Symbol, Timestamp};
+use std::fmt::{Display, Formatter};
+
+/// Generic display for anything that could be converted to `IonType`.
+fn display_type<T>(value: T, f: &mut Formatter<'_>) -> std::fmt::Result
+where
+    T: Into<IonType>,
+{
+    let ion_type: IonType = value.into();
+    write!(f, "{}", ion_type)
+}
+
+/// Subset of [`IonType`] that are strictly the non-null, non-container types.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub enum ScalarType {
+    Bool,
+    Int,
+    Float,
+    Decimal,
+    Timestamp,
+    String,
+    Symbol,
+    Blob,
+    Clob,
+}
+
+impl Display for ScalarType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        display_type(*self, f)
+    }
+}
+
+impl TryFrom<IonType> for ScalarType {
+    type Error = IonError;
+
+    fn try_from(value: IonType) -> Result<Self, Self::Error> {
+        use ScalarType::*;
+        match value {
+            IonType::Bool => Ok(Bool),
+            IonType::Int => Ok(Int),
+            IonType::Float => Ok(Float),
+            IonType::Decimal => Ok(Decimal),
+            IonType::Timestamp => Ok(Timestamp),
+            IonType::String => Ok(String),
+            IonType::Symbol => Ok(Symbol),
+            IonType::Blob => Ok(Blob),
+            IonType::Clob => Ok(Clob),
+            _ => illegal_operation(format!("{} type is not a scalar", value)),
+        }
+    }
+}
+
+/// Subset of [`IonType`] that are strictly the container types.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub enum ContainerType {
+    SExp,
+    List,
+    Struct,
+}
```

Comment:
I am going to cut an issue for this--I am currently not quite comfortable doing it only because we have a `ContainerType` lurking that has something equivalent to *Datagram* in it.

Reply:
Cut #549 for this.